src/storage/src/hummock/compactor/mod.rs

```diff
@@ -671,35 +671,55 @@ impl Compactor {
                 .filter_key_extractor_manager
                 .update_bloom_filter_avg_size(sst_info.file_size as usize, bloom_filter_size);
             let sst_size = sst_info.file_size;
+            let sst_id = sst_info.id;
             ssts.push(sst_info);
 
             let tracker_cloned = task_progress_tracker.clone();
-            upload_join_handles.push(upload_join_handle.and_then(|_| async move {
-                if let Some(tracker) = tracker_cloned {
-                    tracker.inc_ssts_uploaded();
-                }
-                Ok(())
-            }));
-
-            if self.context.is_share_buffer_compact {
-                self.context
-                    .stats
-                    .shared_buffer_to_sstable_size
-                    .observe(sst_size as _);
-            } else {
-                self.context.stats.compaction_upload_sst_counts.inc();
-            }
+            let context_cloned = self.context.clone();
+            upload_join_handles.push(upload_join_handle.and_then(
+                move |upload_result| async move {
+                    if upload_result.is_ok() {
+                        if let Some(tracker) = tracker_cloned {
+                            tracker.inc_ssts_uploaded();
+                        }
+                        if context_cloned.is_share_buffer_compact {
+                            context_cloned
+                                .stats
+                                .shared_buffer_to_sstable_size
+                                .observe(sst_size as _);
+                        } else {
+                            context_cloned.stats.compaction_upload_sst_counts.inc();
+                        }
+                    }
+                    Ok((sst_id, upload_result))
+                },
+            ));
         }
 
+        // Check if there are any failed uploads. Report all of those SSTs.
+        let mut failed_uploads = vec![];
         try_join_all(upload_join_handles)
```

Comment:
We may turn `upload_join_handles` into `upload_join_handles.map_err(HummockError::sstable_upload_error)` so that `try_join_all` will complete with error when any of the concurrent join handles returns an error without waiting other join handle to finish.

Reply:
Wouldn't it be more informative if we can report all failed ssts?

Reply:
I think we only care about whether an compaction task succeeds or not and the cause of the failed task. Any failed SST upload would cause a compaction task to fail, and we only need to report such cause to failure instead of all failed SSTs. New SSTs is not visible to any component in our system until the compaction task succeeds. When any of the SST upload failed, the compaction task is bound to fail, and then the upload result of other SSTs makes no difference to the compaction task and is not some useful information. Instead of collecting these less useful information, I will prefer to fail fast to save some time, and in the future we can even cancel the join handle so that we may save some unnecessary IO.

BTW, if we really care about the upload result of other SSTs when one SST failed to upload, we can add log in the uploader.

Reply:
Yes, let's go with log in the uploader + fail fast.